### PR TITLE
update-ci-testing

### PR DESCRIPTION
**Pull request by itziakos**
Add more recent python version on ci testing

### CODE REVIEW

**Comment by itziakos** [comment]
@sjagoe I am trying to get the github tests to run, but I think I am doing something wrong. Can you have have a look when you have some spare time?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -17,118 +17,36 @@ env:
...
-
+    - uses: actions/checkout@v4
+    - name: Setup python
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python }}
+    - name: Install
+      run: |
+        pip install -e .
+        pip install -r "test-requirements.txt"
```
Comment (sjagoe) [inline .github/workflows/test.yml:35]: Looks like the requirements file is named `test_requirements`

Comment (itziakos) [reply .github/workflows/test.yml:35]: bad copy-pasta. Sorry for the noise

**Comment by sjagoe** [comment]
I can run tests locally if I run from a different directory. If I run from the git checkout directory, I get the same errors :thinking:

**Comment by itziakos** [comment]
> I can run tests locally if I run from a different directory. If I run from the git checkout directory, I get the same errors 🤔

I suspect we get some weird PYTHONPATH interference. I am going to have a closer look

**Comment by sjagoe** [comment]
> I can run tests locally if I run from a different directory

I think that's what tox was doing as well; running from a different location.

**Comment by sjagoe** [comment]
when `haas` is a directory in the current working directory, running  `python -m unittest discover haas` will interpret `haas` as a directory name and add it to `sys.path`

I think it's a unittest limitation and that's partly why tox was used to run in a different directory. We can just create an empty dir and use it as `working_directory` in the run tests step

`python -m haas haas` runs tests as expected even when run from the level above the package :)

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -38,7 +38,7 @@ jobs:
     - name: Run tests
       continue-on-error: false
       run: |
-        cd ci
+        cd .github/workflows
```
Comment (sjagoe) [inline .github/workflows/test.yml:41]: This can be `working-directory: .github/workflows` above the `run:`

**Comment by itziakos** [comment]
> when `haas` is a directory in the current working directory, running `python -m unittest discover haas` will interpret `haas` as a directory name and add it to `sys.path`
> 
> I think it's a unittest limitation and that's partly why tox was used to run in a different directory. We can just create an empty dir and use it as `working_directory` in the run tests step
> 
> `python -m haas haas` runs tests as expected even when run from the level above the package :)

Done :P

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -17,118 +17,39 @@ env:
...
-            python: "3.10.0-alpha.2"
-            toxenv: "cp310"
-            continue-on-error: true
-          - os: windows-2019
-            python: "3.10.0-alpha.2"
-            toxenv: "cp310"
-            continue-on-error: true
-
+        python-version: [3.7, 3.8, 3.9, "3.10", "3.11", '3.12', 'pypy2', 'pypy3']
+        os: [windows-2019, ubuntu-20.4, macos-11]
```
Comment (sjagoe) [inline .github/workflows/test.yml:24]:
```suggestion
        os: [windows-2019, ubuntu-20.04, macos-11]
```

Comment (sjagoe) [reply .github/workflows/test.yml:24]:
```suggestion
        os: [windows-2019, ubuntu-22.04, macos-11]
```

**Comment by itziakos** [comment]
@sjagoe it looks that testing is now working as expected. I will try to fix the real errors next and revisit the setup at then of the PR work.

**Comment by sjagoe** [comment]
It looks like `assertRegexpMatches` was removed in 3.12 and we need to change to `assertRegex`.

The other macos failures have me really confused

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -17,118 +17,39 @@ env:
...
+    - name: Install
+      run: |
+        pip install -e .
+        pip install -r "test_requirements.txt"
+        pip --version
+    - name: Run tests
+      continue-on-error: false
+      working-directory: .github
+      run: |
+        python -m unittest discover haas
```
Comment (sjagoe) [inline .github/workflows/test.yml:42]: Apparently if we use the `--top .` option to `unittest discover`, we don't need to change the working directory:

```
python -m unittest discover --top . haas
```

We may also want to add coverage again with

```
coverage run --branch -m unittest discover -v --top . haas
```

**Pull request merged**